[IOSP-514] [CNSMR-3272] Technical Documentation for Stevenson and CRP

AliSoftware opened the pull request:
Tickets:
* https://babylonpartners.atlassian.net/browse/IOSP-514
* https://babylonpartners.atlassian.net/browse/CNSMR-3272

### Why?

We need detailed documentation about the implementation details of the CRP bot, especially so that if someone needs to update the code it's easily manageable

### How?

* Made the README.md of the Stevenson repo be more focused about the Stevenson framework (it being an OpenSource project and library, the goal of the README is to be read by anyone wanting to use the Stevenson.framework to build their own JIRA/CI/GH bot too)
* Provided high level documentation in `BabylonCommands.md` for the specific Babylon commands implemented in the App target of our bot. It's mainly to tell in this OSS repo how Stevenson is used in our use case, and mostly contains links to the playbook doc where all the Babylon-Specific doc for the bot resides

_(Note: this PR started initially by documenting the CRP technical implementation in that repo. We later decided to move that tech impl doc in the playbook, making that PR way smaller now, and not so related to CRP doc anymore, but still contains some nice cleanup and linking to actual CRP doc in playbook)_

Companion PR in the playbook focusing on Babylon-specific and CRP implementation [is here](https://github.com/babylonhealth/ios-playbook/pull/313)

### PR checklist

* [x] I've assigned this PR to myself

With :heart:

Commit:
Reorganize some files in subfolders for project readability

Commit:
Update README.md to be more general

And not Babylon-specific

Commit:
Summary of supported Babylon Commands implemented in our Babylon Bot

Commit:
Initial doc for CRP Implementation details

Commit:
CRP Impl details doc

Commit:
typo

Commit:
FixedVersionReport impl precisions

viorel15 commented on Documentation/BabylonCommands.md:
```diff
@@ -0,0 +1,110 @@
+
+This command creates a "CRP Ticket" in our JIRA, which is a JIRA ticket gathering all the tickets that will be part of a release to the Stores in order to validate approval for that release. The creation and approval of this ticket to make a new release is part of our SSDLC. 
+
+For more information about the CRP, visit the dedicated page in your Playbook.
```
Can you please link the page?

AliSoftware replied:
Sure 👍 

It was not created when I wrote that PR, but now the PR to create the page exists in playbook (https://github.com/babylonhealth/ios-playbook/pull/313) so I should be able to link it

```suggestion
For more information about the CRP, visit [the dedicated page in your Playbook](https://github.com/babylonhealth/ios-playbook/blob/master/Cookbook/Technical-Documents/CRP-Bot.md).
```
_(link will obviously not work until ios-playbook#313 is merged to master)_

viorel15 commented on Documentation/BabylonCommands.md:
```diff
@@ -0,0 +1,110 @@
+
+* If `branch:` is not provided, defaults to running the lane on `develop`
+* `target:` is only used by some lanes in Fastlane, like `appcenter` or `testflight`. In most cases it's not needed
+* Any potential `<other_options>` depend on each lane. Check our Fastlane to see which are available for each. Examples include `version:` for `appcenter` and `testflight` lanes
```
full examples would help understanding the command. Especially for people unfamiliar with this kind of flow.
I think full examples would be useful for commands below too

AliSoftware replied:
I didn't want those to be too much detailed in here – as I was trying to focus on technical details and implementation, which any _usage documentation_ of babylon-specific commands should be documented in the playbook, not here.

tbh I was even hesitant to keep that `BabylonCommands.md` document here in the repo – its goal was just to describe in a technical POV examples of commands that we could (and did) implement with the `Stevenson.framework`, for anyone else in the OSS community that would be interested to use those as examples/inspiration to implement their own bot using Stevenson-the-framework

I could totally add examples in here but I feel it would make more sense to keep detailed usage documentation in our playbook. Maybe we could even try to remove some paragraphs in that `BabylonCommands.md` doc to really only focus on the implementation details rather than usage description, but it's hard to talk about implementation without explaining even a little the expected thing we are trying to parse 😉

viorel15 replied:
I thought this work was strictly about the process ticket CNSMR-3272. That doc should focus more on the usage rather then implementation. It would be great if you can keep this document that you have just created and have a shorter version focused on the usage. What do you think?

AliSoftware replied:
Yeah the separation as to where document things is not always clear indeed.

We agreed that the documentation should ideally be split in:
 - General Technical documentation about the framework itself (intended for users of the OSS framework) ➡️ in the repo itself
 - Anything specific to Babylon (especially process and specific commands) ➡️ in the playbook

Maybe I can just move this `BabylonCommands.md` doc in the playbook after all then since it's about the usage and specific to commands implemented for Babylon (and not doc for the bot framework intended for OSS users). But **we already have documentation in the playbook describing those commands [here](https://github.com/babylonhealth/ios-playbook/blob/master/Cookbook/Technical-Documents/SlackCIIntegration.md) though** (as well as in the playbook doc about release process) so it would not make sense to duplicate the documentation.

Maybe I will just delete that `BabylonCommands.md` doc entirely then since it doesn't really talk about implementation details after all (and usage doc already exists in playbook), wdyt?

AliSoftware replied:
@viorel15 as discussed in Slack, cleaned up the `BabylonCommands.md` to only focus on implementation details, and links to usage in playbook otherwise.

Still kept the document since we needed a place especially to document the environment variables used by our bot instance (being technical documentation for the bot, I believe that belonged more in this repo than the playbook), but otherwise simplified the doc to only point to relevant places.

ilyapuchka replied:
@AliSoftware if that's the only purpose of this doc that's left (to me it feels it does not add much more clarity about implementation details that we already have in other docs or in code except describing the structure of the project) wouldn't be better to document environment variables in code instead of a separate document and then remove this doc? We can have a small section in the main README dedicated to the structure of the project and references that we have in a playbook.

AliSoftware commented on Documentation/BabylonCommands.md:
```diff
@@ -0,0 +1,110 @@
+
+We have configured our GitHub repo to send the webhook to the URL of our bot, which means our bot will handle every new comment made to any issue or PR and be able to act on it.
+
+This route processes comments starting with `@ios-babylon-bot` (discarding any other kind of comments); it then runs `fastlane` to invoke the mentioned lane on the PR's branch in CI.
```
Actually it invokes a CI workflow, not a fastlane lane.
```suggestion
This route processes comments starting with `@ios-babylon-bot` (discarding any other kind of comments); it then invokes the workflow (whose name is provided in the comment) on the PR's branch in CircleCI.
```

AliSoftware commented on Documentation/BabylonCommands.md:
````diff
@@ -0,0 +1,110 @@
+This route processes comments starting with `@ios-babylon-bot` (discarding any other kind of comments); it then runs `fastlane` to invoke the mentioned lane on the PR's branch in CI.
+
+```
+@ios-bot-babylon <lane_name>
````
```suggestion
@ios-bot-babylon <workflow_name>
```

AliSoftware commented on Documentation/BabylonCommands.md:
````diff
@@ -0,0 +1,110 @@
+@ios-bot-babylon <lane_name>
+```
+
+This will simply run lane `<lane_name>` in CI on the head branch of the PR this comment was made on.
````
```suggestion
This will simply run the workflow `<workflow_name>` on CircleCI, on the head branch of the PR this comment was made on.
```

Commit:
Apply suggestions from code review

Commit:
Add link to playbook doc

AliSoftware commented on Documentation/BabylonCommands.md:
```diff
@@ -0,0 +1,110 @@
+
+This command creates a "CRP Ticket" in our JIRA, which is a JIRA ticket gathering all the tickets that will be part of a release to the Stores in order to validate approval for that release. The creation and approval of this ticket to make a new release is part of our SSDLC. 
+
+For more information about the CRP, visit [the dedicated page in your Playbook](https://github.com/babylonhealth/ios-playbook/blob/master/Cookbook/Technical-Documents/CRP-Bot.md).
```
PR pending for this future document: https://github.com/babylonhealth/ios-playbook/pull/313
(So not yet available in master… until the PR in playbook is merged)

Link to current version in branch in the meantime: https://github.com/babylonhealth/ios-playbook/blob/olivier/IOSP-514/Cookbook/Technical-Documents/CRP-Bot.md

Commit:
Reworded BabylonCommands to be more implementation-focused

Commit:
wording

ilyapuchka commented on Documentation/BabylonCommands.md:
```diff
@@ -0,0 +1,50 @@
+
+ENV Var | Description
+--------|--------------
+`SLACK_TOKEN` | Your access token for the Slack API of your team's Slack
```
imo it would be better to comment that in code, if the purpose of these environment variables is not clear, not to go out of sync. Readme already mentions environment variables so I'd leave this whole section out.

ilyapuchka commented on Documentation/BabylonCommands.md:
```diff
@@ -0,0 +1,50 @@
+* Usage and supported commands are documented [in our playbook here](https://github.com/babylonhealth/ios-playbook/blob/master/Cookbook/Technical-Documents/SlackCIIntegration.md).
+* The commands are implemented/defined in `SlackCommand+Fastlane.swift` and then handled in `SlackService.swift` by `func handle(command:on:)`.
+
+Implementation simply consists of parsing the slack command parameters, then invoking `CircleCIService.runlane` with the right arguments.
```
Is this really helpful and not already clear from code? I'd also avoid using "simply"/"just"/etc.

```suggestion
Implementation consists of parsing the slack command parameters, then invoking `CircleCIService.runlane` with the right arguments.
```

ilyapuchka commented on Documentation/BabylonCommands.md:
```diff
@@ -0,0 +1,50 @@
+Allows us to use a slash command in our Slack to trigger CI workflows.
+
+* Usage and supported commands are documented [in our playbook here](https://github.com/babylonhealth/ios-playbook/blob/master/Cookbook/Technical-Documents/SlackCIIntegration.md).
+* The commands are implemented/defined in `SlackCommand+Fastlane.swift` and then handled in `SlackService.swift` by `func handle(command:on:)`.
```
This is pretty clear from code already, isn't it?

```suggestion
* The commands are implemented/defined in `SlackCommand+Fastlane.swift`.
```

ilyapuchka commented on Documentation/BabylonCommands.md:
```diff
@@ -0,0 +1,50 @@
+## Responding to a GitHub comment to trigger a CI pipeline
+
+* Usage documentation can be found [here](https://github.com/babylonhealth/ios-playbook/blob/master/Cookbook/Technical-Documents/SlackCIIntegration.md)
+* Implementation simply consists of a _route_ implemented in `routes.swift` and invoking `GitHubService.issueComment(on:ci:)` (in `GitHubService+IssueComment.swift`).  
```
IMO that's also pretty clear from the code. What we can mention here is that comments handling is implemented as a separate route in the app.

```suggestion
* Implementation consists of a _route_ implemented in `routes.swift` and invoking `GitHubService.issueComment(on:ci:)` (in `GitHubService+IssueComment.swift`).  
```

ilyapuchka commented on Documentation/BabylonCommands.md:
```diff
@@ -0,0 +1,50 @@
+
+* Usage documentation can be found [here](https://github.com/babylonhealth/ios-playbook/blob/master/Cookbook/Technical-Documents/SlackCIIntegration.md)
+* Implementation simply consists of a _route_ implemented in `routes.swift` and invoking `GitHubService.issueComment(on:ci:)` (in `GitHubService+IssueComment.swift`).  
+  The method parses the webhook request headers and the comment body, then calls `CircleCIService.runLane` or `CircleCIService.runPipeline` appropriately to trigger the right CI workflow.
```
```suggestion
  The method parses the webhook request headers and the comment body and then triggers appropriate CI workflow (that runs a single lane or a pipeline).
```

ilyapuchka commented on README.md:
```diff
@@ -111,12 +107,11 @@ If you need to create a new Slack command:
  2. Click on "Create New Command"
    * Fill in the slash command (e.g. `/foo`)
-   * Enter `https://<appname>.herokuapp.com/<command>` as the request URL, replacing `<appname>` with the name of your Heroku app instance (e.g. `stevenson-bot`) and `<command>` by the command name (e.g. `foo`)
+   * Enter the URL this command will trigger – e.g. if your instance will be hosted on Heroku  it will look like `https://<appname>.herokuapp.com/<command>`, replacing `<appname>` with the name of your Heroku app instance (e.g. `stevenson-bot`) and `<command>` by the command name (e.g. `foo`)
```
```suggestion
   * Enter the URL this command will trigger – e.g. if your instance will be hosted on Heroku it will look like `https://<appname>.herokuapp.com/<command>`, replacing `<appname>` with the name of your Heroku app instance (e.g. `stevenson-bot`) and `<command>` by the command name (e.g. `foo`)
```

ilyapuchka commented on README.md:
```diff
@@ -180,7 +171,8 @@ To set the aforementioned environment variables with the real values on Heroku:
 ## 📖 Documentation
 
-Visit the Vapor web framework's [documentation](http://docs.vapor.codes) for instructions on how to use Vapor.
+* To read the documentation about the Stevenson's instance we use at Babylon and the commands that we implemented in our usage of Stevenson, see [this page of the documentation](Documentation/BabylonCommands.md)
```
```suggestion
* To read the documentation about the Stevenson's instance we use at Babylon and the commands that we implemented to support our processes, see [this page of the documentation](Documentation/BabylonCommands.md)
```

ilyapuchka commented on Documentation/BabylonCommands.md:
```diff
@@ -0,0 +1,50 @@
+This Vapor project consists actually of two projects:
+
+ * The Stevenson framework, aimed to be reused so you can create your own bots interacting with Slack, GitHub, CircleCI and JIRA yourselves
+ * The Stevenson Babylon app, which is the instance of the Bot we use at Babylon, based on the Stevenson framework, and which provides us with the Slack commands we need
```
```suggestion
 * The Stevenson Babylon app, which is the instance of the Bot we use at Babylon, based on the Stevenson framework, and which provides us with commands we need
```

ilyapuchka commented on Documentation/CRP.md:
```diff
@@ -0,0 +1,103 @@
+# CRP Implementation Details
```
Considering that you mention "Anything specific to Babylon (especially process and specific commands) ➡️ in the playbook" wouldn't be better to have it in a playbook?

As of the content of this doc I thought it would be more about how this command should work and why (business rules) rather than implementation details with methods names etc. I feel this would me more valuable and implementation details documentation belongs more to the code comments.

AliSoftware replied:
Business rules are documented in the other PR in the playbook, which indeed document the usage.

I can move implementation details to the playbook too if it makes more sense. Would also allow to reduce that document by removing anything that's trivial when just reading the code

(My initial understanding of the AC was that everything about technical implementation should be documented in repo, but I agree that code is pretty much documented or clear already)

AliSoftware commented:
@dmcrodrigues What are your thoughts about this? You initially wanted this documentation so that anyone could maintain the CRP if I wasn't around, but I mostly agree with the feedback given here that it feels strange to have it here – even though it's what we initially stated in the AC.

_Usage_ documentation is already available in the Playbook (https://github.com/babylonhealth/ios-playbook/pull/313). So should this _technical_ documentation be:

 - in this repo as it is in this PR because it's about _technical impl_
 - moved in playbook too because it's kinda babylon-specific, even if it's in the OSS repo
 - removed altogether (and potentially replaced with just some additional doc comments in the code where needed), because most of the implementation details feels quite trivial and self-explanatory from the code anyway?

viorel15 reviewed (approved):
👍

danspndl commented on Documentation/BabylonCommands.md:
```diff
@@ -0,0 +1,50 @@
+# Babylon's Stevenson Commands
+
+This Vapor project consists actually of two projects:
```
[nit]
```suggestion
This Vapor project actually consists of two projects:
```

danspndl commented on Documentation/BabylonCommands.md:
```diff
@@ -0,0 +1,50 @@
+
+# 🕹 Commands implemented in the Babylon Bot
+
+The Stevenson instance used at Babylon, whose code can be found in the `App/` directory, mainly consist of 3 main core features
```
```suggestion
The Stevenson instance used at Babylon, whose code can be found in the `App/` directory, consist of 3 main core features
```

danspndl commented on README.md:
```diff
@@ -1,7 +1,7 @@
 
-`Stevenson` is a Vapor framework designed to build integrations between Slack apps, Github, JIRA and CI services (CircleCI).
-This project also contains implementation of the Slack app used by Babylon iOS team (if you want to know more about how our team works checkout our [playbook](https://github.com/babylonhealth/ios-playbook))
+`Stevenson` is a Vapor framework designed to build integrations between Slack apps, Github, JIRA and CI services (CircleCI).  
```
👀 
```suggestion
`Stevenson` is a Vapor framework designed to build integrations between Slack apps, GitHub, JIRA and CI services (CircleCI).
```

danspndl commented on README.md:
```diff
@@ -21,8 +21,9 @@ import Stevenson
 
-`Stevenson` comes with implementation of Slack [slash commands](https://api.slack.com/slash-commands), GitHub, JIRA and CircleCI APIs. At the moment it does not implement complete set of these APIs but only provides bare minimum required for the functionality of the app. 
-To create these services use corresponding type constructors providing required values. It's advised but not required to store these values in the environment variables:
+`Stevenson` comes with implementation of Slack [slash commands](https://api.slack.com/slash-commands), GitHub, JIRA and CircleCI APIs. At the moment it does not implement complete set of these APIs but only provides bare minimum required for the functionality of the app.
```
```suggestion
`Stevenson` comes with implementation of Slack [slash commands](https://api.slack.com/slash-commands), for GitHub, JIRA and CircleCI APIs. At the moment, it does not implement complete set of these APIs but only provides bare minimum required for the functionality of the app.
```

dmcrodrigues commented:
@AliSoftware @ilyapuchka I think overall the documentation that we have here, especially for CRP, is quite valuable and something that helps understanding a bit better how it works and everything involved, we can argue that code should give us more or less the same information but this follows an order which is easier to follow to anyone outside of this specific context in my opinion. Regarding the place where it belongs, for CRP in specific we probably should move it to playbook I agree, is quite specific to Babylon so playbooks feels a better place to host it.

The rest seems more a nice to have but not as fundamental I would say but in any case I don't see a problem in having it. The ENV vars maybe is something that we don't need to document directly but the rest seems fine to have. Makes sense?

Commit:
Apply suggestions from code review

Co-Authored-By: Daniel Spindelbauer <dan.spindelbauer@babylonhealth.com>
Co-Authored-By: Ilya Puchka <ilya.puchka@babylonhealth.com>

Commit:
Merge branch 'master' into olivier/IOSP-514-CNSMR-3272

Commit:
Moved CRP.md to playbook – See https://github.com/babylonhealth/ios-playbook/pull/313

Commit:
Remove list of Env Vars in doc

AliSoftware commented:
@dmcrodrigues @ilyapuchka @viorel15 Updated the PR:

* Moved `CRP.md` doc to the playbook (see https://github.com/babylonhealth/ios-playbook/pull/313)
* Updated links to that document in `README` and `BabylonCommands.md`
* Moved `BabylonCommands.md` back to top level
* Removed the list of env vars from `BabylonCommands.md`; the paragraph still exists but just references where in the code they are declared – and that you need to add them to your scheme to debug the bot

viorel15 reviewed (approved):
👍

NSMyself reviewed (approved):
👌

dmcrodrigues commented on BabylonCommands.md:
```diff
@@ -0,0 +1,45 @@
+
+The bot instance of Stevenson we use at Babylon uses environment variables to instantiate and configure the various services (see [README.md](../README.md#supported-services) and `configure.swift`).
+
+The list of all those environment variables we use can be found [in `app.swift` (see `extension Environment { … }`)](https://github.com/babylonhealth/Stevenson/blob/master/Sources/App/app.swift#L13-L21)
```
👍
